packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js

```diff
@@ -0,0 +1,81 @@
+// @ts-check
+import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava.js';
+import { AmountMath as m, makeIssuerKit } from '@agoric/ertp';
+import { natSafeMath } from '@agoric/zoe/src/contractSupport/index.js';
+import fc from 'fast-check';
+import { balancesToReachRatio } from '../../../src/vpool-xyk-amm/addLiquidity.js';
+
+const { add, multiply } = natSafeMath;
+
+const { brand: brandX } = makeIssuerKit('X');
+const { brand: brandY } = makeIssuerKit('Y');
+
+const oneB = 1_000_000_000n;
+// const dec18 = 1_000_000_000_000_000_000n; // 18 decimals as used in ETH
+
+const arbPoolX = fc
+  .bigUint({ max: oneB })
+  .map(value => m.make(brandX, 100n + value));
+const arbPoolY = fc
+  .bigUint({ min: 100n, max: oneB })
+  .map(value => m.make(brandY, 100n + value));
+const arbGiveX = fc
+  .bigUint({ min: 100n, max: oneB })
+  .map(value => m.make(brandX, value));
+const arbGiveY = fc
+  .bigUint({ min: 100n, max: oneB })
+  .map(value => m.make(brandY, value));
+
+function withinEpsilon(numer, denom, epsilon) {
+  return (
+    Number(denom) * (1 - epsilon) <= numer &&
+    numer <= Number(denom) * (1 + epsilon)
+  );
+}
+
+test('balancesToReachRatio calculations are to spec', t => {
+  fc.assert(
+    fc.property(
+      fc.record({
+        poolX: arbPoolX,
+        poolY: arbPoolY,
+        giveX: arbGiveX,
+        giveY: arbGiveY,
+      }),
+      ({ poolX, poolY, giveX, giveY }) => {
+        const { targetX, targetY } = balancesToReachRatio(
+          poolX,
+          poolY,
+          giveX,
+          giveY,
+        );
+
+        if (targetX.value === 0n && targetY.value === 0n) {
+          t.pass();
+        } else {
+          const targetProduct = Number(multiply(targetX.value, targetY.value));
+          const withinRange = withinEpsilon(
+            targetProduct,
+            Number(multiply(poolX.value, poolY.value)),
+            0.05,
+          );
```

Comment:
One of my goals in doing property-based testing (that's the term of art, btw; not parameter-based) was to avoid thinking about floating point arithmetic. Again, the spec (JSDoc) hasn't been updated to say anything about conversion to floating point, so it's hard for me to evaluate whether this test is correct.

Reply:
The test now only converts to `Number` in order to create a ratio in the parameter for the call to `imbalancedRequest()` so the ratio doesn't coerce to zero if it's less than 1. `imbalancedRequest()`  uses  `Number(startK)` when multiplying by desiredRatio (a fraction that is less than 1.)